packages/ember-routing/lib/system/router_state.js

```diff
@@ -11,11 +11,8 @@ export default EmberObject.extend({
     let state = this.routerJsState;
     if (!this.routerJs.isActiveIntent(routeName, models, null, state)) { return false; }
 
-    let emptyQueryParams = isEmpty(Object.keys(queryParams));
-
-    if (queryParamsMustMatch && !emptyQueryParams) {
-      let visibleQueryParams = {};
-      assign(visibleQueryParams, queryParams);
+    if (queryParamsMustMatch && Object.keys(queryParams).length > 0) {
```

Comment:
This still loops keys three times, once to check length, once to clone and once to shallow equal on clone, seems like this could use more cleanup